### DISCUSSION TITLE
Need help with RC extraction

### DISCUSSION THREAD

**Discussion by daipvd** [The-OpenROAD-Project/OpenROAD#5500] [General]
hi Everyone,
This is first time I use openRoad to extract. I tried to modify my def file like the picture below but I got "_Def parser has encountered an error in file my_contact.def at line 94, on token RECT_". Please help to show me what wrong and how can I fix it?

![image](https://github.com/user-attachments/assets/effb22b3-de20-4d21-8afb-28760ce65ac5)

Thank you,
dai

**Comment by maliberty** [comment]
That is invalid syntax per the LEF/DEF manual.  Regular nets need symbolic routes not rects.

**Reply by daipvd** [reply]
Thank you, @maliberty 
If use "routed" statement like below then it mean I can not change width of routed net(width of net was default in tech lef), right? 
I'm not sure if I'm correct, but I reviewed the syntax documentation of the .def file, and it seems to support the RECT statement. As I understand it, OpenRoad might not support reading the RECT statement, right?

![image](https://github.com/user-attachments/assets/b4d79db1-3c53-4999-b82d-2d0e273c0260)

Once again, I want to thank you for your help.

dai

**Reply by maliberty** [reply]
What are you trying to accomplish?  Your understand of DEF seems to be inaccurate and this is not an OR limitation.

**Reply by daipvd** [reply]
Hi @maliberty,

Thank you for your response. I am trying to create a DEF file for a net route(from GDS file) where the width used on that net varies (please see the attached image below). I'm not sure how to describe that net in the DEF file so that OpenRoad can extract it. Could you please advise?
![image](https://github.com/user-attachments/assets/9ea114dd-ee96-4979-8c2d-2a5d09983f11)

Best regards,
dai

**Reply by maliberty** [reply]
If you are using gds why not extract in a tool like Magic?  In LEF you would use a width WRONGDIRECTION rule to specify the wrong way jogs need to be wider.

**Reply by daipvd** [reply]
Thank you for your suggestion, I'm not familiar with Magic now but I will give it a try. If you can give me some instructions to describe the net in DEF file now?

Thank you very much.
dai

**Reply by maliberty** [reply]
If you update the LEF then you just need the sequence of points in DEF and no rects.

**Reply by daipvd** [reply]
hi @maliberty ,

Here is my run_ext.tcl file:
"
read_lef tech.tlef
read_lef library.lef
read_def -order_wires my_contact.def

set extRules RCmax.rules
set spef_out my_contact.spef

source via_resistance.tcl

define_process_corner -ext_model_index 0 X
extract_parasitics -ext_model_file $extRules
write_spef $spef_out

exit
"
I change M1 in tech.tlef file like your instructions:
![image](https://github.com/user-attachments/assets/3bf5e59f-2827-414f-85d5-d752e71eebe4)

If just need the sequence of point (points in center line of the wire) so how does OR know which segment of wire is wider?
"
  NETS 1 ;
  - INX
    ( PIN INX ) ( x2 Z ) ( x1 Z )  + ROUTED M1 ( 3120 995 ) ( 2220 995 ) ( 2220 395 ) ( 1360 395 ) ;  
  END NETS
"

![image](https://github.com/user-attachments/assets/a8c83443-dd48-423b-ba6c-236dffac192e)

Thank you very much,
dai

**Reply by maliberty** [reply]
It knows the preferred direction of the layer and can use the appropriate rule for the segments.

**Reply by daipvd** [reply]
Hi maliberty,

Base on tech lef file I think OR knows the preferred direction.
As you see in the picture I have three widths used in this net (0.1um, 0.2um(verical) and 0.15um). There is no information about the widths in DEF file and other input files so I'm still considering how does it know? 

Thank you so much for your patience in explaining this to me.  I truly appreciate it!

dai

**Reply by maliberty** [reply]
The width of the wrong way segment is expressed through the LEF rule, not through DEF.  If it were in DEF every wrong way segment would require notation.

**Reply by daipvd** [reply]
Can you show me how to do notation for a wrong way segment in DEF file?
It is great if you can give me an example.

Thanks you,
dai

**Reply by maliberty** [reply]
Its expressed in LEF not DEF so I can't give an example in DEF.

**Reply by daipvd** [reply]
So can you give me an example in LEF file?

Thank a lot,
dai

**Reply by maliberty** [reply]
It looks like you already have the change in your "I change M1 in tech.tlef file like your instructions:"

**Reply by daipvd** [reply]
oh, I understood, so after I change like that in tech lef file do I need to re-generate RC rule file?

Thank you,
dai

**Reply by maliberty** [reply]
No it should be fine.

**Reply by daipvd** [reply]
> If you are using gds why not extract in a tool like Magic? In LEF you would use a width WRONGDIRECTION rule to specify the wrong way jogs need to be wider.

It is a realy good suggestion @maliberty, I tried on magic and it is exactly what I need. I don't need a converting gds to def step more.

Thank you very much !
dai
